test/parallel/test-fs-promises.js

```diff
@@ -336,6 +336,38 @@ async function executeOnHandle(dest, func) {
       await unlink(newFile);
     }
 
+    // Testing readdir recursivly lists both files and directories
+    {
+      const fileStructure = [
+        'a',
+        ['foo', ['b','c']],
+        ['bar', [['fuzz', ['d']], 'e']]
+      ]
+      // Create file structure
+      async function createFiles (path, fileStructure) {
+        for (const fileOrDir of fileStructure) {
+          if (typeof fileOrDir === 'string') {
+            await (await open(`${path}/${fileOrDir}`, 'w')).close();
+          } else {
+            await mkdir(`${path}/${fileOrDir[0]}`);
+            await createFiles(`${path}/${fileOrDir[0]}`, fileOrDir[1]);
+          }
+        }
+      }
+
+      await createFiles(tmpDir, fileStructure)
+
+      const expected = [ 'a', 'bar/e', 'bar/fuzz/d', 'foo/b', 'foo/c' ]
+
+      // resolve all at once
+      assert.deepStrictEqual(expected, (await readdir(tmpDir, { recursive: true })).sort());
```

Comment:
I'm not sure this should return an array (so maybe the functionality should be implemented in a new method or within `opendir`), as it could potentially be a huge result and take a long time to return.

Reply:
I was just coming here to comment the same, that `opendir` might be the better method to add the recursive option to.

Would be curious what @cjihrig thinks, who just switched `cp` to using `opendir`.

Reply:
Thanks for the ping. I agree with this point. I mentioned that in https://github.com/nodejs/node/issues/34992#issuecomment-685976091 as well.

It's already been called out in this PR, but using `opendir()` will make it unnatural for this API to return an array. IMO, that makes it awkward to integrate the proposed change into `readdir()`. You could probably make the streaming API support a recursive option though.

Reply:
i know this is a very old comment, but to continue development I am resolving it. Please comment this again on the latest changes if it still stands true.